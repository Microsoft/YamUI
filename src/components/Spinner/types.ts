/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum SpinnerColor {
  /**
   * Light theme.
   */
  LIGHT = 'light',
  /**
   * Light theme spinner color with gray text.
   */
  METADATA = 'metadata',
  /**
   * Dark theme.
   */
  DARK = 'dark',
}
export enum SpinnerSize {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
