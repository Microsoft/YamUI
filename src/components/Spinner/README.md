# Spinner

A `Spinner` is an outline of a circle which animates around itself indicating to the user that things are processing. It is shown when we're unsure how long a task will take.

## Properties

| Prop | Type | Description |
| ---- | ---- | ----------- |
| ariaLabel | string | Alternative label for screen readers. |
| ariaLive | string | Politeness for label update announcement. |
| color | SpinnerColor | Color of the spinner. |
| label | string | Label displayed besides/below the spinner. |
| size | SpinnerSize | Size of spinner. |

## Examples

### Default spinner

```js
<Spinner />
```

### Large spinner

```js
<Spinner size={ SpinnerSize.LARGE } />
```

### Spinner on a dark background

```js
<Spinner color={SpinnerColor.DARK} />
```

### Spinner with label

```js
<Spinner label="Loading..." />
```
