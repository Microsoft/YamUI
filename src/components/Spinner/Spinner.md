A `Spinner` is an outline of a circle which animates around itself indicating to the user that things are processing. It is shown when we're unsure how long a task will take.

### Examples

Spinners without label:

```js
const { SpinnerColor, SpinnerSize } = require('.');

<table>
  <thead>
    <tr>
      <th style={{ width: '100px' }}>L</th>
      <th style={{ width: '100px' }}>M</th>
      <th style={{ width: '100px' }}>S</th>
      <th style={{ width: '100px' }}>XS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Spinner size={SpinnerSize.LARGE} />
      </td>
      <td>
        <Spinner size={SpinnerSize.MEDIUM} />
      </td>
      <td>
        <Spinner size={SpinnerSize.SMALL} />
      </td>
      <td>
        <Spinner size={SpinnerSize.XSMALL} />
      </td>
    </tr>
    <tr style={{ backgroundColor: '#386cbb' }}>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.LARGE} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.MEDIUM} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.SMALL} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.XSMALL} />
      </td>
    </tr>
  </tbody>
</table>
```

Spinners with label:

```js
const { SpinnerColor, SpinnerSize } = require('.');

<table>
  <thead>
    <tr>
      <th style={{ width: '100px' }}>L</th>
      <th style={{ width: '100px' }}>M</th>
      <th style={{ width: '100px' }}>S</th>
      <th style={{ width: '100px' }}>XS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Spinner size={SpinnerSize.LARGE} label="Loading" />
      </td>
      <td>
        <Spinner size={SpinnerSize.MEDIUM} label="Loading" />
      </td>
      <td>
        <Spinner size={SpinnerSize.SMALL} label="Loading" />
      </td>
      <td>
        <Spinner size={SpinnerSize.XSMALL} label="Loading" />
      </td>
    </tr>
    <tr style={{ backgroundColor: '#386cbb' }}>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.LARGE} label="Loading" />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.MEDIUM} label="Loading" />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.SMALL} label="Loading" />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.XSMALL} label="Loading" />
      </td>
    </tr>
  </tbody>
</table>
```
