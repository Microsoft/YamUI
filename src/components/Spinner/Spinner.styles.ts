/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { addAlpha, getTheme } from '../../util/colors';
import { mergeStyleSets } from '@uifabric/styling';
import { memoizeFunction } from '@uifabric/utilities';
import { SpinnerColor, SpinnerSize } from './types';

export interface SpinnerClassNames {
  root: string;
  label: string;
  circle: string;
}

const labelMargins = {
  [SpinnerSize.XSMALL]: '-1px 0 0 6px',
  [SpinnerSize.SMALL]: '-1px 0 0 8px',
  [SpinnerSize.MEDIUM]: '-2px 0 0 8px',
  [SpinnerSize.LARGE]: '4px 0 0',
};

export const getClassNames = memoizeFunction(
  (color: SpinnerColor, size: SpinnerSize, isCentered: boolean): SpinnerClassNames => {
    const theme = getTheme();
    const circleColor = color === SpinnerColor.DARK ? theme.palette.white : theme.semanticColors.link;

    return mergeStyleSets({
      root: {
        display: 'flex',
        alignItems: 'center',
        flexDirection: size === SpinnerSize.LARGE ? 'column' : 'row',
        justifyContent: isCentered ? 'center' : undefined,
      },
      circle: {
        selectors: {
          '> .ms-Spinner-circle': {
            border: `2px solid ${addAlpha(circleColor, 30)}`,
            borderTopColor: circleColor,
          },
        },
      },
      label: {
        margin: labelMargins[size],
      },
    });
  },
);
