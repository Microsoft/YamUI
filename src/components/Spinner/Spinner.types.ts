/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';

export enum SpinnerColor {
  /**
   * Light theme.
   */
  LIGHT = 'light',
  /**
   * Light theme spinner color with gray text.
   */
  METADATA = 'metadata',
  /**
   * Dark theme.
   */
  DARK = 'dark',
}

export enum SpinnerSize {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}

export interface SpinnerProps extends BaseComponentProps {
  /**
   * Text displayed besides/below the spinner.
   */
  text: string;

  /**
   * Whether to hide text, and only render it for screenreaders.
   * @default false
   */
  hideText?: boolean;

  /**
   * Color of the spinner.
   * @default SpinnerColor.LIGHT
   */
  color?: SpinnerColor;

  /**
   * Size of the spinner.
   * @default SpinnerSize.MEDIUM
   */
  size?: SpinnerSize;

  /**
   * Will center the Spinner horizontally within its container.
   */
  isCentered?: boolean;
}
