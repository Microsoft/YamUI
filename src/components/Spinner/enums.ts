/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum SpinnerSize {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
