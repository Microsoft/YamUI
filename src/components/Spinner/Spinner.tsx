/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { CustomizableComponentProps, defaultTheme, customizable } from '../Customizer';
import {
  ISpinnerProps as FabricSpinnerProps,
  Spinner as FabricSpinner,
  SpinnerSize as FabricSpinnerSize,
} from 'office-ui-fabric-react/lib/Spinner';
import Text, { TextColor, TextSize } from '../Text';
import ScreenreaderText from '../ScreenreaderText';
import { SpinnerProps, SpinnerColor, SpinnerSize } from './Spinner.types';
import { join } from '../../util/classNames';
import { getClassNames, getFabricSpinnerStyles } from './Spinner.styles';

const labelColors = {
  [SpinnerColor.LIGHT]: TextColor.PRIMARY,
  [SpinnerColor.METADATA]: TextColor.METADATA,
  [SpinnerColor.DARK]: TextColor.WHITE,
};

const labelSizes = {
  [SpinnerSize.LARGE]: TextSize.MEDIUM_SUB,
  [SpinnerSize.MEDIUM]: TextSize.MEDIUM_SUB,
  [SpinnerSize.SMALL]: TextSize.MEDIUM_SUB,
  [SpinnerSize.XSMALL]: TextSize.SMALL,
};

const fabricSpinnerSizes = {
  [SpinnerSize.LARGE]: FabricSpinnerSize.large,
  [SpinnerSize.MEDIUM]: FabricSpinnerSize.medium,
  [SpinnerSize.SMALL]: FabricSpinnerSize.small,
  [SpinnerSize.XSMALL]: FabricSpinnerSize.xSmall,
};

export class Spinner extends React.Component<SpinnerProps & CustomizableComponentProps> {
  public render() {
    const {
      theme = defaultTheme,
      text,
      hideText,
      color = SpinnerColor.LIGHT,
      size = SpinnerSize.MEDIUM,
      isCentered = false,
    } = this.props;

    const textSize = labelSizes[size];
    const classNames = getClassNames({ size, isCentered, theme, textSize });

    const label = hideText ? (
      <ScreenreaderText>{text}</ScreenreaderText>
    ) : (
      <Text className={classNames.label} color={labelColors[color]} size={textSize}>
        {text}
      </Text>
    );

    return (
      <div className={join(['y-spinner', classNames.root])}>
        <FabricSpinner {...this.getSpinnerProps()} />
        {label}
      </div>
    );
  }

  private getSpinnerProps(): FabricSpinnerProps {
    const { color = SpinnerColor.LIGHT, size, theme = defaultTheme } = this.props;

    return {
      size: fabricSpinnerSizes[size as SpinnerSize],
      styles: getFabricSpinnerStyles({ color, theme }),
    };
  }
}

/**
 * A `Spinner` is an outline of a circle which animates around itself indicating to the user that
 * things are processing. It is shown when we're unsure how long a task will take.
 */
@customizable('Spinner', ['theme'])
export default class CustomizableSpinner extends Spinner {}
