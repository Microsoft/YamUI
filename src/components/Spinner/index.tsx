/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { Enum } from 'typescript-string-enums';
import {
  ISpinnerProps as FabricSpinnerProps,
  Spinner as FabricSpinner,
  SpinnerSize as FabricSpinnerSize,
} from 'office-ui-fabric-react/lib/Spinner';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import Text, { TextProps, TextColor, TextSize } from '../Text';
import { SpinnerSize } from '../../util/enums/spinner';
import './spinner.css';

export { SpinnerSize };

const SizeMap = {
  [SpinnerSize.LARGE]: FabricSpinnerSize.large,
  [SpinnerSize.MEDIUM]: FabricSpinnerSize.medium,
  [SpinnerSize.SMALL]: FabricSpinnerSize.small,
  [SpinnerSize.XSMALL]: FabricSpinnerSize.xSmall,
};

export const SpinnerColor = Enum({
  /**
   * Light theme.
   */
  LIGHT: 'light',
  /**
   * Dark theme.
   */
  DARK: 'dark',
});
export type SpinnerColor = Enum<typeof SpinnerColor>;

export interface SpinnerProps extends BaseComponentProps {
  /**
   * Alternative label for screen readers.
   */
  ariaLabel?: string;

  /**
   * Politeness for label update announcement.
   * @default polite
   */
  ariaLive?: FabricSpinnerProps['ariaLive'];

  /**
   * Color of the spinner.
   * @default SpinnerColor.LIGHT
   */
  color?: SpinnerColor;

  /**
   * Label displayed besides/below the spinner.
   */
  label?: string;

  /**
   * Size of the spinner.
   * @default SpinnerSize.MEDIUM
   */
  size?: SpinnerSize;
}

export default class Spinner extends React.PureComponent<SpinnerProps, {}> {
  static defaultProps: Partial<SpinnerProps> = {
    ariaLive: 'polite',
    color: SpinnerColor.LIGHT,
    size: SpinnerSize.MEDIUM,
  };

  render() {
    const { label } = this.props;

    return (
      <div className={this.getClasses()}>
        <FabricSpinner {...this.getSpinnerProps()} />
        {label && <Text {...this.getLabelProps()}>{label}</Text>}
      </div>
    );
  }

  private getSpinnerProps(): FabricSpinnerProps {
    return {
      ariaLabel: this.props.ariaLabel || this.props.label,
      ariaLive: this.props.ariaLive,
      size: SizeMap[this.props.size as string],
      className: 'y-spinner--circle',
    };
  }

  private getLabelProps(): TextProps {
    return {
      color: this.props.color === SpinnerColor.DARK ? TextColor.WHITE : TextColor.PRIMARY,
      size: this.props.size === SpinnerSize.XSMALL ? TextSize.SMALL : TextSize.MEDIUM_SUB,
      className: 'y-spinner--label',
    };
  }

  private getClasses() {
    const { props } = this;
    const classes: string[] = [
      'y-spinner',
      `y-spinner__color-${props.color}`,
      `y-spinner__size-${props.size}`,
    ];
    if (this.props.className) {
      classes.push(this.props.className);
    }
    return classes.join(' ');
  }
}
