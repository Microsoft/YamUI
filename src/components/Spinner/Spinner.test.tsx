/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Spinner, { SpinnerProps, SpinnerColor, SpinnerSize } from '.';

describe('<Spinner />', () => {
  let component: ShallowWrapper<SpinnerProps, {}>;

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(
        <Spinner className="TEST_CLASSNAME" text="Loading" />,
      );
    });

    it('includes that className', () => {
      expect(component.hasClass('TEST_CLASSNAME')).toBe(true);
    });

    it('still includes its base className', () => {
      expect(component.hasClass('y-spinner')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with visible text', () => {
    beforeEach(() => {
      component = shallow(
        <Spinner text="Loading" />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with invisible text', () => {
    beforeEach(() => {
      component = shallow(
        <Spinner text="Loading" hideText={true} />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with a color', () => {
    beforeEach(() => {
      component = shallow(
        <Spinner text="Loading" color={SpinnerColor.DARK} />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with a size', () => {
    beforeEach(() => {
      component = shallow(
        <Spinner text="Loading" size={SpinnerSize.XSMALL} />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
