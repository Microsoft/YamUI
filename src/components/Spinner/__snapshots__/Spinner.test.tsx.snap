// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> when centered matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        justify-content: center;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0.8rem;
          margin-right: 0;
          margin-top: -1px;
        }
    color="primary"
    size="mediumSub"
  >
    Loading
  </CustomizedText>
</div>
`;

exports[`<Spinner /> with additional className matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0.8rem;
          margin-right: 0;
          margin-top: -1px;
        }
    color="primary"
    size="mediumSub"
  >
    Loading
  </CustomizedText>
</div>
`;

exports[`<Spinner /> with dark color matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(255, 255, 255, 0.3)",
          "borderTopColor": "#ffffff",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0.8rem;
          margin-right: 0;
          margin-top: -1px;
        }
    color="white"
    size="mediumSub"
  >
    Loading
  </CustomizedText>
</div>
`;

exports[`<Spinner /> with invisible text matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <ScreenreaderText>
    Loading
  </ScreenreaderText>
</div>
`;

exports[`<Spinner /> with large size matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: column;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    size={3}
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0;
          margin-right: 0;
          margin-top: 0.4rem;
        }
    color="primary"
    size="mediumSub"
  >
    Loading
  </CustomizedText>
</div>
`;

exports[`<Spinner /> with metadata color matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0.8rem;
          margin-right: 0;
          margin-top: -1px;
        }
    color="metadata"
    size="mediumSub"
  >
    Loading
  </CustomizedText>
</div>
`;

exports[`<Spinner /> with visible text matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        min-height: 2.0rem;
      }
>
  <StyledSpinnerBase
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0.8rem;
          margin-right: 0;
          margin-top: -1px;
        }
    color="primary"
    size="mediumSub"
  >
    Loading
  </CustomizedText>
</div>
`;

exports[`<Spinner /> with xSmall size matches its snapshot 1`] = `
<div
  className=
      y-spinner
      {
        align-items: center;
        display: flex;
        flex-direction: row;
        min-height: 1.6rem;
      }
>
  <StyledSpinnerBase
    size={0}
    styles={
      Object {
        "circle": Object {
          "border": "2px solid rgba(56, 108, 187, 0.3)",
          "borderTopColor": "#386cbb",
        },
      }
    }
  />
  <CustomizedText
    className=

        {
          margin-bottom: 0;
          margin-left: 0.6rem;
          margin-right: 0;
          margin-top: 0;
        }
    color="primary"
    size="small"
  >
    Loading
  </CustomizedText>
</div>
`;
